### PR TITLE
Add hipdnn convolution support

### PR DESCRIPTION
**Pull request by zjgarvey**
Implements forward and backward convolution (2D/3D) through the hipDNN frontend graph API, providing an alternative to the MIOpen backend on ROCm.

- **Shared utilities**: Extract `createTensorAttributes()` into `ATen/hipdnn/Utils.h` for reuse across hipDNN ops (BatchNorm, Conv)
- **Graph-cached convolution**: Forward (fprop), backward-data (dgrad), and backward-weight (wgrad) via hipDNN frontend graphs with a thread-local LRU cache (`ParamsLRUCache<K,V>`) to amortize `graph->build()` cost
- **Dispatch integration**: New `ConvBackend::Hipdnn` and `ConvBackend::HipdnnTranspose` variants wired through backend selection, memory format selection, forward/backward switches, and Python enum exposure. hipDNN takes priority over MIOpen when `torch.backends.hipdnn.enabled` is `True`
- **Bias fusion**: Forward conv fuses bias via a pointwise ADD node in the graph, avoiding a separate `output.add_()` call
- **Transposed convolution**: Implemented as dgrad (forward) / fprop (backward-input) / wgrad (backward-weight), with bias applied separately (for now) since dgrad + pointwise graphs aren't supported by any hipDNN backends currently.
- **Grouped/depthwise support**: hipDNN infers group count from tensor dimensions; explicit output dims are set on dgrad/wgrad graphs so grouping is resolved correctly

### CODE REVIEW

File: aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+          DEFAULT_LIMIT,
+          " entries.");
+    }
+    return DEFAULT_LIMIT;
+  }();
+  return limit;
+}
+
+template <typename KeyType>
+struct HipdnnGraphCache {
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp]: Factor out some op independent LRU cache utils.

File: aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+    IntArrayRef dilation) {
+
+  auto inputType = getHipdnnDataType(grad_output);
+  auto graph = std::make_shared<hipdnn_frontend::graph::Graph>();
+  // No set_intermediate_data_type needed: single-op graph has no virtual tensors.
+  graph->set_io_data_type(inputType)
+      .set_compute_data_type(hipdnn_frontend::DataType::FLOAT);
+
+  auto dy_attr = createTensorAttributes(grad_output);
+  dy_attr->set_uid(UID_INPUT);
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp:242]: This is sloppy. Why use named struct for something where the names don't align?

File: aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp [hunk 3]
```diff
@@ -0,0 +1,659 @@
...
+    cache->update(key, std::move(entry));
+    cached = cache->find(key);
+  }
+
+  std::unordered_map<int64_t, void*> variantPack;
+  variantPack[cached->input_uid] = grad_output.data_ptr();
+  variantPack[cached->weight_uid] = weight.data_ptr();
+  variantPack[cached->output_uid] = grad_input.data_ptr();
+
+  auto workspace = at::empty({cached->workspace_size}, grad_output.options().dtype(at::kByte));
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp:423]: How is it guaranteed that this empty space is on the correct device?

File: aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+  auto memory_format = hipdnn_conv_suggest_memory_format(input_t, weight_t);
+  auto input_c = input_t.contiguous(memory_format);
+  auto weight_c = weight_t.contiguous(memory_format);
+
+  auto trans_output_size = conv_input_size(
+      input_c.sizes(), weight_c.sizes(), padding, output_padding, stride, dilation, groups);
+  auto output = at::empty(trans_output_size, input_c.options(), memory_format);
+
+  // Transposed conv forward is dgrad (no bias fusion — dgrad graph has no bias path)
+  runHipdnnConvDgrad(input_c, weight_c, output,
```
Comment (zjgarvey) [inline aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp]: Is the `output_padding` inferred from the `trans_output_size`? I suppose so. We should check this doesn't break for unusual values.

File: aten/src/ATen/native/Convolution.cpp [hunk 5]
```diff
@@ -1629,6 +1656,18 @@ at::Tensor _convolution(
       output = output.view(calc_output_size(input, weight, params));
       break;
     }
+    case ConvBackend::Hipdnn:
+      check_input_same_type_as_parameters(input, weight, bias);
+      output = at::hipdnn_convolution(
+          input.contiguous(backend_memory_format), weight, bias, params.padding, params.stride,
```
Comment (zjgarvey) [inline aten/src/ATen/native/Convolution.cpp]: Why are we passing contiguous here AND inside this function? Same with the transpose conv below.

File: test/test_hipdnn_conv.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+import unittest
+
+import torch
+import torch.nn.functional as F
+from torch.testing._internal.common_cuda import TEST_HIPDNN
+from torch.testing._internal.common_utils import run_tests, TestCase
+
+
+@unittest.skipIf(not TEST_HIPDNN, "hipDNN not available")
+class TestHipdnnConvolution(TestCase):
```
Comment (zjgarvey) [inline test/test_hipdnn_conv.py:12]: We can't expect allclose here unless sizes are small. Will need to take a closer look here.

**Review by zjgarvey — COMMENTED** [review]
Self-review

File: aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+Tensor hipdnn_convolution(
+    const Tensor& input_t,
+    const Tensor& weight_t,
+    const std::optional<Tensor>& bias_opt,
+    IntArrayRef padding,
+    IntArrayRef stride,
+    IntArrayRef dilation,
+    int64_t groups,
+    bool benchmark,
+    bool deterministic) {
```
Comment (BrianHarrisonAMD) [inline aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp:488]: Not sure if deterministic / benchmark are in-scope for this first phase, but for hipDNN we can support these right now with a bit of extra work.

For deterministic, there is a deterministic MIOpen engine that can be selected which will limit solutions to the deterministic ones (matching with existing integration).

For benchmark, there exists a benchmarking knob inside the MIOpen plugins that can be set to allow the tuning phase to occur. We would need to add some code that gets the knobs for the selected engine, and set's benchmarking if it's available.

File: aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+#include <ATen/native/utils/ParamsLRUCache.h>
+#include <c10/util/env.h>
+#include <c10/util/irange.h>
+
+namespace at::native {
+
+// ---------------------------------------------------------------------------
+// Cache key: captures everything that determines graph topology
+// ---------------------------------------------------------------------------
+constexpr int hipdnn_max_dim = 3;
```
Comment (dnikolaev-amd) [inline aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp:40]: MIOPEN_DIM_MAX = 5 already defined and used
https://github.com/ROCm/pytorch/blob/2befc15f29ccda07111d7a82bd4f9490e8a9d091/aten/src/ATen/native/Normalization.cpp#L67
https://github.com/ROCm/pytorch/blob/2befc15f29ccda07111d7a82bd4f9490e8a9d091/aten/src/ATen/native/Convolution.cpp#L90
https://github.com/ROCm/pytorch/blob/f6a72d1464e0e5f99cae8d3196c7990d0f3bcdff/aten/src/ATen/miopen/Descriptors.cpp#L28

File: aten/src/ATen/hipdnn/Handle.cpp
Comment (mousdahl-amd) [inline aten/src/ATen/hipdnn/Handle.cpp]: Do we need `hipdnn_frontend::detail` here?  That namespace is for internal implementation details.

File: aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp (hunk 8 above)
Comment (zjgarvey) [reply aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp:40]: I'd like to leave this separate, since I don't want to use any of these named constants from other source files. 

E.g., hipdnn might have other backends which support more dim combinations than miopen.

**Comment by zjgarvey** [comment]
I think I need a rebase, one sec.

File: aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp (hunk 7 above)
Comment (zjgarvey) [reply aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp:488]: I could add this in if you think it is high priority for conv.

Comment (zjgarvey) [reply aten/src/ATen/native/hipdnn/Conv_hipdnn.cpp:488]: For now, let me plumb these through to the graph construction functions and raise an error for deterministic and raise a warning for benchmark.

File: aten/src/ATen/native/Convolution.cpp [hunk 9]
```diff
@@ -492,6 +494,15 @@ struct ConvParams {
            && !(groups > 1 && is_dilated()) // MIOpen currently does not support dilation with groups of size > 1
            ;
   }
+  bool use_hipdnn(const at::Tensor& input, const at::Tensor& weight) const {
+    if (!at::globalContext().userEnabledHipdnn()) return false;
+    if (!detail::getCUDAHooks().compiledWithHipDNN()) return false;
+    if (!input.is_cuda()) return false;
+    auto dtype = input.scalar_type();
```
Comment (mousdahl-amd) [inline aten/src/ATen/native/Convolution.cpp:499]: These datatype / dimension checks are interesting to me.  I understand wanting a fast path to skip if hipDNN doesn't support certain datatypes / tensor dimensions, but this could very easily change going forward.  I wonder if there's a way we can check in with hipDNN to see if it's got applicable engines instead.  I just want to avoid a maintenance headache if possible.

The problem with that is it may be weightier than you want to do.

**Review by mousdahl-amd — COMMENTED** [review]
Nothing major jumped out at me.

**Comment by zjgarvey** [comment]
I'm landing this so we can rebase hipdnn_develop into modular conv / batch norm changes.
